Get block notifications API

lock9 opened the pull request:
Closes #3779.

It works but I don't know if this is the best way to do it.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+		return nil, respErr
+	}
+
+	block, err := s.chain.GetBlock(hash)
```
I think we need some `GetTrimmedBlock` or alike, this one does extract all transactions and we don't need them (only hash is needed for `GetAppExecResults()`). But it's an optimization in its essence, current code will work fine.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+	for _, tx := range block.Transactions {
+		aers, err := s.chain.GetAppExecResults(tx.Hash(), trigger.Application)
+		if err != nil {
+			continue
```
Return internal error? This means some DB inconsistency and the result can be incorrect.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+		}
+	}
+
+	// Also check for notifications from the block itself (PostPersist)
```
You need to keep the sequence correct and check for `trigger.OnPersist` _before_ digging into transaction list.

AnnaShaleva replied:
We may additionally extend `neorpc.NotificaitonFilter` with trigger parameter, it may be useful in some cases, what do you think?

roman-khimov replied:
What are the use cases? People interested in block-level things (and not a lot happens there) probably pull block applogs directly and otherwise it's all `Application`.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+	}
+
+	// Also check for notifications from the block itself (PostPersist)
+	aers, err := s.chain.GetAppExecResults(block.Hash(), trigger.Application)
```
`trigger.PostPersist` here.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+	aers, err := s.chain.GetAppExecResults(block.Hash(), trigger.Application)
+	if err == nil && len(aers) > 0 {
+		aer := aers[0]
+		if aer.VMState == vmstate.Halt {
```
Maybe this code can be generalized in some way (helper function?), it's the same for blocks/transactions.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+}
+
+// testComparator is a helper type for notification filtering
+type testComparator struct {
```
You can move these out into a separate file (along with the logic for appending to `notifications`.

AnnaShaleva replied:
Also, I'd say that `testComparator/testContainer` is suitable for testing code, but not for the production code. Let's rename it to something more appropriate.

roman-khimov commented on pkg/services/rpcsrv/server_test.go:
```diff
@@ -2274,6 +2274,43 @@ var rpcTestCases = map[string][]rpcTestCase{
+			params: `["` + genesisBlockHash + `", {"contract":"` + testContractHashLE + `", "name":"Transfer"}]`,
+			result: func(e *executor) any { return []state.ContainedNotificationEvent{} },
+			check: func(t *testing.T, e *executor, acc any) {
+				res, ok := acc.([]state.ContainedNotificationEvent)
```
Please check for expected contents as well.

roman-khimov commented:
RPC client needs to be updated to support this as well.

AnnaShaleva commented on pkg/services/rpcsrv/server_test.go:
```diff
@@ -2274,6 +2274,43 @@ var rpcTestCases = map[string][]rpcTestCase{
+		},
+		{
+			name:   "positive with filter",
+			params: `["` + genesisBlockHash + `", {"contract":"` + testContractHashLE + `", "name":"Transfer"}]`,
```
Genesis block doesn't contain notifications from testContract. Let's use native Gas or Neo hash as a filter.

lock9 replied:
Updated: Now using Neo token hash

AnnaShaleva commented on pkg/services/rpcsrv/server.go:
```diff
@@ -219,6 +220,7 @@ var rpcHandlers = map[string]func(*Server, params.Params) (any, *neorpc.Error){
 	"getblockhash":                 (*Server).getBlockHash,
 	"getblockheader":               (*Server).getBlockHeader,
 	"getblockheadercount":          (*Server).getBlockHeaderCount,
+	"getblocknotifications":        (*Server).getBlockNotifications,
```
This extension should be documented at `docs/rpc.md`, see `### Extensions` section.

AnnaShaleva replied:
This extension should be supported at the RPC client side (`pkg/rpcclient/rpc.go`).

AnnaShaleva reviewed (commented):
A useful extension.

AnnaShaleva commented:
Please, ensure testing jobs are passing. Currently, unit-tests, project linter and DCO checks are failing.

lock9 commented:
Hey guys, got some busy days ahead. I'll come back to this in one or two weeks max. Sorry for the delay

lock9 commented:
PR Updated. 

Changes:
- Added GetTrimmedBlock and TrimmedBlock
  - Note: There's already the concept of 'trimmed block' and this name may not be ideal
 - Added a new struct "BlockNotifications", grouping notifications by trigger type
 - Updated the tests to use the neo token hash
 - Updated the tests to check the notification content
- Changed the name of some functions/types
- Moved reusable logic to the utils file
- Added the new endpoint to the rpc docs file
- Updated the client class to support the new endpoint

Still pending:
- RPC Client tests

About trigger filters: I don't know/don't have an opinion.

lock9 commented on pkg/core/block/block.go:
```diff
@@ -241,3 +254,94 @@ func (b *Block) ToStackItem() stackitem.Item {
+	return nil
+}
+
+func (b *TrimmedBlock) ToStackItem() stackitem.Item {
```
is this necessary?

AnnaShaleva commented on pkg/services/rpcsrv/util.go:
```diff
@@ -18,3 +24,46 @@ func checkInt32(i int) error {
 	return nil
 }
+
+func (c notificationComparatorFilter) EventID() neorpc.EventID {
```
All exported methods require a comment.

AnnaShaleva replied:
s/notificationComparatorFilter/notificationEventComparator

AnnaShaleva commented on pkg/services/rpcsrv/util.go:
```diff
@@ -18,3 +24,46 @@ func checkInt32(i int) error {
 }
+
+func (c notificationComparatorFilter) EventID() neorpc.EventID {
+	return c.id
```
`c.id` field is not needed, remove it. `notificatoinEventComparator` works only with `ContainedNotificatoinEvent`, hence you may return `neorpc.NotificationEventID` here.

AnnaShaleva commented on pkg/services/rpcsrv/util.go:
```diff
@@ -18,3 +24,46 @@ func checkInt32(i int) error {
+			Container:         container,
+			NotificationEvent: evt,
+		}
+		if filter == nil || rpcevent.Matches(&notificationComparatorFilter{
```
`filter == nil` check is excessive. `rpcevent.Matche` perfectly handle this case.

AnnaShaleva commented:
> About trigger filters: I don't know/don't have an opinion.

Seems that we don't need it for now, so nothing should be done wrt this question.

AnnaShaleva commented:
Please, format commits according to https://github.com/nspcc-dev/.github/blob/master/git.md#commits. See https://github.com/nspcc-dev/neo-go/pull/3787/commits for example.

Also, please ensure your commits have valid signed-off-by signature.

roman-khimov commented on pkg/neorpc/result/block_notifications.go:
```diff
@@ -0,0 +1,12 @@
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+)
+
+// BlockNotifications represents notifications from a block organized by trigger type.
+type BlockNotifications struct {
+	PrePersistNotifications  []state.ContainedNotificationEvent `json:"prepersist,omitempty"`
+	TxNotifications          []state.ContainedNotificationEvent `json:"transactions,omitempty"`
+	PostPersistNotifications []state.ContainedNotificationEvent `json:"postpersist,omitempty"`
```
`[]ContainedNotificationEvent` is sufficient to me, it will be ordered natually, it has triggers inside to separate pre/post. We better not introduce additional types if we can.

AnnaShaleva replied:
> it has triggers inside

It doesn't, that's the case.

roman-khimov replied:
Ah, OK. Likely we don't care for notifications, if something happened it happened. Container is here anyway.

roman-khimov replied:
Ah, OK. Likely we don't care for notifications, if something happened it happened. Container is here anyway.

roman-khimov replied:
You can still drop `Notifications` from all of the field names, it's excessive. Actually following trigger names makes total sense here, OnPersist/Application/PostPersist like in https://github.com/nspcc-dev/neo-go/blob/8d728b4ec179e6354c21fc7e2c4f8a7c4df7312e/pkg/smartcontract/trigger/trigger_type.go

lock9 commented on pkg/core/block/block.go:
```diff
@@ -241,3 +254,94 @@ func (b *Block) ToStackItem() stackitem.Item {
+	return block, br.Err
+}
+
+func (b TrimmedBlock) MarshalJSON() ([]byte, error) {
```
At first, I considered adding a get trimmed block endpoint. I also had some issues to debug the RPC endpoints (it doesn't stop on the debugger). I had an  error that I didn't know where it was coming from. Later I realized that it was due to the notification object serialization, not the block. The same for the trimmed block. I didn't know if that was being used (now I know). 

I also got confused because the tests 'were passing', when in practice, they weren't. I 'suspected' that this was related to some inner serilization / deserialization that I wasn't aware of. That's why I included these methods.  I'll remove this and other parts that aren't being used.

roman-khimov commented on pkg/core/dao/dao.go:
```diff
@@ -439,6 +439,21 @@ func (dao *Simple) getBlock(key []byte) (*block.Block, error) {
 }
 
+// GetTrimmedBlock returns a block with only the header and transaction hashes.
+func (dao *Simple) GetTrimmedBlock(hash util.Uint256) (*block.TrimmedBlock, error) {
```
Why do we need it? `GetBlock()` already does the same thing at the DAO level.

roman-khimov commented on pkg/core/block/block.go:
```diff
@@ -221,7 +235,6 @@ func (b *Block) GetExpectedBlockSizeWithoutTransactions(txCount int) int {
 	return size
 }
 
-// ToStackItem converts Block to stackitem.Item.
```
Still relevant.

roman-khimov commented on pkg/core/dao/dao_test.go:
```diff
@@ -123,6 +124,55 @@ func TestPutGetBlock(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestGetTrimmedBlock(t *testing.T) {
```
Leftover, fails to test.

roman-khimov commented on pkg/core/blockchain.go:
```diff
@@ -3149,3 +3149,8 @@ func (bc *Blockchain) GetStoragePrice() int64 {
 }
+
+// GetTrimmedBlock returns a block with only the header and transaction hashes.
+func (bc *Blockchain) GetTrimmedBlock(hash util.Uint256) (*block.TrimmedBlock, error) {
```
Needs to be fixed, fails to build.

roman-khimov commented on pkg/neorpc/result/block_notifications.go:
```diff
@@ -0,0 +1,12 @@
+
+// BlockNotifications represents notifications from a block organized by trigger type.
+type BlockNotifications struct {
+	PrePersistNotifications  []state.ContainedNotificationEvent `json:"prepersist,omitempty"`
```
I think we better have `onpersist` instead of `prepersist` in JSON, while `pre` is technically more correct here we have `on` used in trigger, method and many other places, so it'd be a bit more consistent this way.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3212,51 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+	notifications := &result.BlockNotifications{}
+
+	aers, err := s.chain.GetAppExecResults(block.Hash(), trigger.OnPersist)
+	if err == nil && len(aers) > 0 {
```
Still relevant.

roman-khimov commented on pkg/services/rpcsrv/server.go:
```diff
@@ -3202,3 +3212,51 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
+	}
+
+	aers, err = s.chain.GetAppExecResults(block.Hash(), trigger.PostPersist)
+	if err == nil && len(aers) > 0 {
```
Also.

roman-khimov commented on pkg/services/rpcsrv/notification_comparator.go:
```diff
@@ -0,0 +1,65 @@
+			NotificationEvent: evt,
+		}
+		if rpcevent.Matches(&notificationEventComparator{
+			filter: *filter,
```
Won't this panic on `nil` filter? Probably something was a bit different when https://github.com/nspcc-dev/neo-go/pull/3781#discussion_r1922358523 was made.

roman-khimov commented:
I can't push to linkd-academy:block-notifications, so this one is replaced with #3805 (which includes changes from this PR), thanks!
